job/dependency_resolver.go

```diff
@@ -70,30 +70,62 @@ func (r *dependencyResolver) resolveInferredDependencies(ctx context.Context, jo
 
 	// get job spec of these destinations and append to current jobSpec
 	for _, depDestination := range jobDependencies {
-		depSpec, depProj, err := projectJobSpecRepo.GetByDestination(ctx, depDestination)
-		if err != nil {
+		projectJobPairs, err := projectJobSpecRepo.GetByDestination(ctx, depDestination)
+		if err != nil || len(projectJobPairs) == 0 {
 			if err == store.ErrResourceNotFound {
 				// should not fail for unknown dependency
 				r.notifyProgress(observer, &EventJobSpecUnknownDependencyUsed{Job: jobSpec.Name, Dependency: depDestination})
 				continue
 			}
 			return jobSpec, errors.Wrap(err, "runtime dependency evaluation failed")
 		}
-
-		// determine the type of dependency
-		dep := models.JobSpecDependency{Job: &depSpec, Project: &depProj}
-		dep.Type = r.getJobSpecDependencyType(dep, projectSpec.Name)
-		jobSpec.Dependencies[depSpec.Name] = dep
+		dep := extractDependency(projectJobPairs, projectSpec)
+		jobSpec.Dependencies[dep.Job.Name] = dep
 	}
 
 	return jobSpec, nil
 }
 
-func (r *dependencyResolver) getJobSpecDependencyType(dependency models.JobSpecDependency, currentJobSpecProject string) models.JobSpecDependencyType {
-	if dependency.Project.Name == currentJobSpecProject {
-		return models.JobSpecDependencyTypeIntra
+// extractDependency extracts tries to find the upstream dependency from multiple matches
+// type of dependency is decided based on if the job belongs to same project or other
+// Note(kushsharma): correct way to do this is by creating a unique destination for each job
+// this will require us to either change the database schema or add some kind of
+// unique literal convention
+func extractDependency(projectJobPairs []store.ProjectJobPair, projectSpec models.ProjectSpec) models.JobSpecDependency {
+	var dep models.JobSpecDependency
+	if len(projectJobPairs) == 1 {
```

Comment:
is this needed? or we can just skip this, and go straight to check if any job within the same project exists or not. that way we can remove the duplication of getting the first element and remove the logic of checking inter dependency I believe.

Reply:
Yes, I agree this will simplify the code but will ignore the fact that this should not happen, we should not have this ambiguity in the first place. If you read the comment on line 110, I wanted to point out that this is not a perfect way to resolve this. Ideally, we should always only get one job for one destination.

Reply:
makes sense 👍